File: backend/app/schemas/audit_discussion_issue.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class AuditDiscussionIssueBase(BaseModel):
    audit_type: str = Field(..., min_length=2, max_length=50)
    discussion_point: str = Field(..., min_length=2, max_length=255)
    default_decision: str = Field(..., min_length=2)
    status: str = Field(default="active", min_length=2, max_length=20)

    @field_validator("audit_type", "discussion_point", "default_decision", "status")
    @classmethod
    def clean_required_text(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("This field is required.")
        return value


class AuditDiscussionIssueCreate(AuditDiscussionIssueBase):
    pass


class AuditDiscussionIssueUpdate(BaseModel):
    audit_type: str | None = Field(default=None, min_length=2, max_length=50)
    discussion_point: str | None = Field(default=None, min_length=2, max_length=255)
    default_decision: str | None = Field(default=None, min_length=2)
    status: str | None = Field(default=None, min_length=2, max_length=20)
    is_active: bool | None = None

    @field_validator("audit_type", "discussion_point", "default_decision", "status")
    @classmethod
    def clean_optional_required_text(cls, value: str | None) -> str | None:
        if value is None:
            return None

        value = value.strip()
        if not value:
            raise ValueError("Field cannot be empty.")
        return value


class AuditDiscussionIssueResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    issue_id: int
    audit_type: str
    discussion_point: str
    default_decision: str
    status: str
    is_active: bool
    created_by: str | None = None
    updated_by: str | None = None
    created_at: datetime
    updated_at: datetime


class AuditDiscussionIssueListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[AuditDiscussionIssueResponse]


class AuditDiscussionIssueMessageResponse(BaseModel):
    message: str
    data: AuditDiscussionIssueResponse | None = None

File: frontend/src/services/auditDiscussionIssue.ts
export type AuditDiscussionIssue = {
  issue_id: number;
  audit_type: string;
  discussion_point: string;
  default_decision: string;
  status: string;
  is_active: boolean;
  created_by: string | null;
  updated_by: string | null;
  created_at: string;
  updated_at: string;
};

export type AuditDiscussionIssueListResponse = {
  total: number;
  page: number;
  page_size: number;
  items: AuditDiscussionIssue[];
};

export type AuditDiscussionIssuePayload = {
  audit_type: string;
  discussion_point: string;
  default_decision: string;
  status: string;
};

export type AuditDiscussionIssueUpdatePayload =
  Partial<AuditDiscussionIssuePayload> & {
    is_active?: boolean;
  };

type ListParams = {
  page: number;
  pageSize: number;
  search?: string;
  isActive?: boolean;
  auditType?: string;
  status?: string;
};

function buildQuery(params: ListParams) {
  const query = new URLSearchParams();

  query.set("page", String(params.page));
  query.set("page_size", String(params.pageSize));
  query.set("sort_by", "issue_id");
  query.set("sort_order", "desc");

  if (params.search) query.set("search", params.search);

  if (typeof params.isActive === "boolean") {
    query.set("is_active", String(params.isActive));
  }

  if (params.auditType) query.set("audit_type", params.auditType);
  if (params.status) query.set("status", params.status);

  return query.toString();
}

async function requestJson<T>(
  url: string,
  options?: RequestInit,
): Promise<T> {
  const response = await fetch(url, {
    ...options,
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
      ...(options?.headers ?? {}),
    },
  });

  if (!response.ok) {
    const error = await response.json().catch(() => null);
    const message =
      error?.detail ||
      error?.message ||
      "Audit Discussion Issue request failed.";

    throw new Error(message);
  }

  return (await response.json()) as T;
}

export async function listAuditDiscussionIssues(params: ListParams) {
  const query = buildQuery(params);

  return requestJson<AuditDiscussionIssueListResponse>(
    `/api/backend/audit-discussion-issues?${query}`,
    {
      method: "GET",
    },
  );
}

export async function createAuditDiscussionIssue(
  payload: AuditDiscussionIssuePayload,
) {
  return requestJson<{ message: string; data: AuditDiscussionIssue }>(
    "/api/backend/audit-discussion-issues",
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
  );
}

export async function updateAuditDiscussionIssue(
  id: number,
  payload: AuditDiscussionIssueUpdatePayload,
) {
  return requestJson<{ message: string; data: AuditDiscussionIssue }>(
    `/api/backend/audit-discussion-issues/${id}`,
    {
      method: "PATCH",
      body: JSON.stringify(payload),
    },
  );
}

export async function deactivateAuditDiscussionIssue(id: number) {
  return requestJson<{ message: string; data: AuditDiscussionIssue }>(
    `/api/backend/audit-discussion-issues/${id}`,
    {
      method: "DELETE",
    },
  );
}

export async function restoreAuditDiscussionIssue(id: number) {
  return requestJson<{ message: string; data: AuditDiscussionIssue }>(
    `/api/backend/audit-discussion-issues/${id}/restore`,
    {
      method: "PATCH",
    },
  );
}

export async function permanentDeleteAuditDiscussionIssue(id: number) {
  return requestJson<{ message: string; data: null }>(
    `/api/backend/audit-discussion-issues/${id}/permanent`,
    {
      method: "DELETE",
    },
  );
}
